Content.Client/Tabletop/TabletopSystem.cs

```diff
@@ -0,0 +1,303 @@
+
+        private float _timePassed; // Time passed since last update sent to the server.
+        private IEntity? _draggedEntity; // Entity being dragged
+        private ScalingViewport? _viewport; // Viewport currently being used
+        private SS14Window? _window; // Current open tabletop window (only allow one at a time)
+        private IEntity? _table; // The table entity of the currently open game session
+
+        public override void Initialize()
+        {
+            CommandBinds.Builder
+                        .Bind(EngineKeyFunctions.Use, new PointerInputCmdHandler(OnUse, false))
+                        .Register<TabletopSystem>();
+
+            SubscribeNetworkEvent<TabletopPlayEvent>(OnTabletopPlay);
+            SubscribeLocalEvent<TabletopDraggableComponent, ComponentHandleState>(HandleComponentState);
+        }
+
+        public override void Update(float frameTime)
+        {
+            // If there is no player entity, return
+            if (_playerManager.LocalPlayer is not { ControlledEntity: { } playerEntity }) return;
+
+            if (StunnedOrNoHands(playerEntity))
+            {
+                StopDragging();
+            }
+
+            if (!CanSeeTable(playerEntity, _table))
+            {
+                StopDragging();
+                _window?.Close();
+                return;
+            }
+
+            // If no entity is being dragged or no viewport is clicked, return
+            if (_draggedEntity == null || _viewport == null) return;
+
+            // Make sure the dragged entity has a draggable component
+            if (!_draggedEntity.TryGetComponent<TabletopDraggableComponent>(out var draggableComponent)) return;
+
+            // If the dragged entity has another dragging player, drop the item
+            // This should happen if the local player is dragging an item, and another player grabs it out of their hand
+            if (draggableComponent.DraggingPlayer != null &&
+                draggableComponent.DraggingPlayer != _playerManager.LocalPlayer?.Session.UserId)
+            {
+                StopDragging(false);
+                return;
+            }
+
+            // Map mouse position to EntityCoordinates
+            var coords = _viewport.ScreenToMap(_inputManager.MouseScreenPosition.Position);
+
+            // Clamp coordinates to viewport
+            var clampedCoords = ClampPositionToViewport(coords, _viewport);
+            if (clampedCoords.Equals(MapCoordinates.Nullspace)) return;
+
+            // Move the entity locally every update
+            _draggedEntity.Transform.LocalPosition = clampedCoords.Position;
+
+            // Increment total time passed
+            _timePassed += frameTime;
+
+            // Only send new position to server when Delay is reached
+            if (_timePassed >= Delay)
+            {
+                RaiseNetworkEvent(new TabletopMoveEvent(_draggedEntity.Uid, clampedCoords));
```

Comment:
Should diff with last position to save on bandwidth flooding

Reply:
I'm not really sure how to do this, I tried this
![image](https://user-images.githubusercontent.com/39844191/128776153-24c08a9c-1806-472c-9b29-1e686be0e13e.png)
but it made it very laggy for the other client for some reason. The position is only sent during dragging, so is it really that important to optimize it?

Reply:
Depends if you want your code to be slow or not.